engine/src/main/java/com/agiletec/aps/system/services/keygenerator/KeyGeneratorManager.java

```diff
@@ -13,74 +13,67 @@
  * @author S.Didaci - E.Santoboni
  */
 public class KeyGeneratorManager extends AbstractService implements IKeyGeneratorManager {
 
-	private static final Logger _logger = LoggerFactory.getLogger(KeyGeneratorManager.class);
-	
-	@Override
-	public void init() throws Exception {
-		this.loadUniqueKey();
-		_logger.debug("{} ready. : last loaded key {}", this.getClass().getName(), _uniqueKeyCurrentValue );
-	}
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private IKeyGeneratorDAO keyGeneratorDao;
+
+    private IKeyGeneratorManagerCacheWrapper cacheWrapper;
+
+    protected IKeyGeneratorDAO getKeyGeneratorDAO() {
+        return keyGeneratorDao;
+    }
+
+    public void setKeyGeneratorDAO(IKeyGeneratorDAO generatorDAO) {
+        this.keyGeneratorDao = generatorDAO;
+    }
 
-	/**
-	 * Estrae la chiave presente nel db.
-	 * Il metodo viene chiamato solo in fase di inizializzazione.
-	 * @throws ApsSystemException
-	 */
-	private void loadUniqueKey() throws ApsSystemException {
-		try {
-			_uniqueKeyCurrentValue = this.getKeyGeneratorDAO().getUniqueKey();
-		} catch (Throwable t) {
-			_logger.error("Error retrieving the unique key", t);
-			//ApsSystemUtils.logThrowable(t, this, "loadUniqueKey","Error retrieving the unique key");
-			throw new ApsSystemException("Error retrieving the unique key", t);
-		}
-	}
+    protected IKeyGeneratorManagerCacheWrapper getCacheWrapper() {
+        return cacheWrapper;
+    }
 
-	/**
-	 * Restituisce la chiave univoca corrente.
-	 * @return La chiave univoca corrente.
-	 * @throws ApsSystemException In caso di errore 
-	 * nell'aggiornamento della chiave corrente.
-	 */
-	public int getUniqueKeyCurrentValue() throws ApsSystemException {
-		++_uniqueKeyCurrentValue;
-		int key = _uniqueKeyCurrentValue;
-		this.updateKey();
-		return key;
-	}
+    public void setCacheWrapper(IKeyGeneratorManagerCacheWrapper cacheWrapper) {
+        this.cacheWrapper = cacheWrapper;
+    }
 
-	private void updateKey() throws ApsSystemException {
-		try {
-			this.getKeyGeneratorDAO().updateKey(_uniqueKeyCurrentValue);
-		} catch (Throwable t) {
-			_logger.error("Error updating the unique key", t);
-			//ApsSystemUtils.logThrowable(t, this, "loadUniqueKey");
-			throw new ApsSystemException("Error updating the unique key", t);
-		}
-	}
+    @Override
+    public void init() throws Exception {
+        this.getCacheWrapper().initCache(this.getKeyGeneratorDAO());
+        logger.debug("{} ready. : last loaded key {}", this.getClass().getName(), this.getCacheWrapper().getUniqueKeyCurrentValue());
+    }
 
-	protected IKeyGeneratorDAO getKeyGeneratorDAO() {
-		return _keyGeneratorDao;
-	}
+    /**
+     * Restituisce la chiave univoca corrente.
+     * @return La chiave univoca corrente.
+     * @throws ApsSystemException In caso di errore 
+     * nell'aggiornamento della chiave corrente.
+     */
+    public synchronized int getUniqueKeyCurrentValue() throws ApsSystemException {

```

Comment:
The update call doesn't need to be synchronized and since it is a DAO and a cache call it would be better if it were outside the block. I think it should be something like

    synchronized (SOME_PRIVATE_STRING) { 
    int key = this.getCacheWrapper().getUniqueKeyCurrentValue() + 1;
    }
    this.updateKey(key);